### PR TITLE
RFC: Add an interfaces manual chapter

### PR DESCRIPTION
**Pull request by mbauman**
Document the iterable, indexable, and abstract array interfaces.

I'm open to suggestions on the tables.  I like the way they summarize the requirements, and I think the iterable and indexable ones work well, but the array table is nearly unreadable since it needs to scroll horizontally.

### CODE REVIEW

**Comment by mbauman** [comment]
There's room at the bottom for someone else to later add information about the Equality interface (`==`/`isequal`/`hash`/`<`/`isless`) and Display interface (`display`/`show`/`print`).  :)

**Comment by tkelman** [comment]
This is great! Should the code blocks be doctests here?

**Comment by mbauman** [comment]
Ah, yes, thank you, they should.

File: doc/manual/interfaces.rst [hunk 1]
```diff
@@ -0,0 +1,223 @@
+.. _man-interfaces:
+
+************
+ Interfaces
+************
+
+A lot of the power and extensibility in Julia comes from a collection of informal interfaces.  By extending few specific methods to work for a custom type, objects of that type not only receive those functionalities, but they are also able to be used in other methods that are written to generically build upon those behaviors.
```
Comment (tkelman) [inline doc/manual/interfaces.rst]: By extending a few`

File: doc/manual/interfaces.rst [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+---------
+
+================================= ======================== ===========================================
+Required methods                                           Brief description
+================================= ======================== ===========================================
+:func:`start(iter) <start>`                                Returns the initial iteration state
+:func:`next(iter, state) <next>`                           Returns the current item and the next state
+:func:`done(iter, state) <done>`                           Tests if there are any items remaining
+**Important optional methods**    **Default definition**   **Brief description**
+:func:`eltype(IterType) <eltype>` ``Any``                  The container's element type
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: I wouldn't use the word container like this, or at least I don't think of all iterables as being containers. Maybe the "The element type of the container or sequence`. Its a really minor point though

File: doc/manual/interfaces.rst [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+:func:`length(iter) <length>`     (*undefined*)            The container's length
+================================= ======================== ===========================================
+
+Sequential iteration is implemented by the methods :func:`start`, :func:`done`, and :func:`next`. Instead of mutating objects as they are iterated over, Julia provides these three methods to keep track of the iteration state externally from the object. The :func:`start(iter)` method returns an initial ``state`` object that gets passed along to :func:`done(iter, state)`, which tests if there are any elements remaining, and :func:`next(iter, state)`, which returns a tuple containing the current element and an updated ``state``. The ``state`` object can be anything, and is generally considered to be an implementation detail private to the iterable object.
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: returns the initial ``state`` for the iterable object ``iter`` that

File: doc/manual/interfaces.rst [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+Any object that has these three methods appropriately defined can be used in a ``for`` loop since the syntax::
+
+    for i in iter   # or  "for i = iter"
+      # body
+    end
+
+is translated into::
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: I think `lowered` is the technical word, but translated probably more familiar to more people

File: doc/manual/interfaces.rst [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+:func:`eltype(IterType) <eltype>` ``Any``                  The container's element type
+:func:`length(iter) <length>`     (*undefined*)            The container's length
+================================= ======================== ===========================================
+
+Sequential iteration is implemented by the methods :func:`start`, :func:`done`, and :func:`next`. Instead of mutating objects as they are iterated over, Julia provides these three methods to keep track of the iteration state externally from the object. The :func:`start(iter)` method returns an initial ``state`` object that gets passed along to :func:`done(iter, state)`, which tests if there are any elements remaining, and :func:`next(iter, state)`, which returns a tuple containing the current element and an updated ``state``. The ``state`` object can be anything, and is generally considered to be an implementation detail private to the iterable object.
+
+Any object that has these three methods appropriately defined can be used in a ``for`` loop since the syntax::
+
+    for i in iter   # or  "for i = iter"
+      # body
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: Four space indent?

File: doc/manual/interfaces.rst [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+    julia> sum(Squares(1803))
+    1955361914
+
+Or even the mean and standard deviation::
+
+    julia> mean(Squares(100)), std(Squares(100))
+    (3383.5,3024.355854282583)
+
+There are a few more methods we can extend to give Julia more information about this iterable collection.  We know that the elements in a ``Squares`` collection will always be ``Int``. By extending the :func:`eltype` method, we can give that information to Julia and help it make more specialized code in the more complicated methods. We also know the number of elements in our collection, so we can extend :func:`length`, too::
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: "``Squares`` sequence"?

File: doc/manual/interfaces.rst [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+Now, when we ask Julia to :func:`collect` all the elements into an array it can preallocate a ``Vector{Int}`` of the right size instead of blindly ``push!``\ ing each element into a ``Vector{Any}``::
+
+    julia> collect(Squares(100))' # transposed to save space
+    1x100 Array{Int64,2}:
+     1  4  9  16  25  36  49  64  81  100  …  9025  9216  9409  9604  9801  10000
+
+While we can rely upon generic implementations, we can also extend specific methods where we know there is a simpler algorithm.  For example, there's a formula to compute the sum of squares, so we can override the generic iterative version with a more performant solution::
+
+    julia> sum(S::Squares) = (n = S.count; return n*(n+1)*(2n+1)÷6)
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: Does this need to be Base.sum?

**Comment by IainNZ** [comment]
Conceptually, what separates these from a "formal" interface? That the type isn't directly annotated with something that says "Iterable"?

File: doc/manual/interfaces.rst [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+Indexing
+--------
+
+====================================== ==================================
+Methods to implement                   Brief description
+====================================== ==================================
+:func:`getindex(X, i) <getindex>`      ``X[i]``, indexed element access
+:func:`setindex!(X, v, i) <setindex!>` ``X[i] = v``, indexed assignment
+:func:`endof(X) <endof>`               The last index, used in ``X[end]``
+====================================== ==================================
```
Comment (IainNZ) [inline doc/manual/interfaces.rst]: What happens with colons, like `Squares[:]`? Should that go here?

Comment (mbauman) [inline doc/manual/interfaces.rst]: Yeah, as I was writing it I briefly considered using that as a motivating segue into the AbstractArray section. Not sure why I didn't end up doing that… I guess I got distracted. `S[:]` now gets lowered to `getindex(S, Colon())`.  So it's entirely up to the type to define the appropriate method.  Definitely worth a mention.

File: doc/manual/interfaces.rst (hunk 7 above)
Comment (mbauman) [inline doc/manual/interfaces.rst]: Keen eye.  It sure does.  `make doctest` found that one, too.

File: doc/manual/interfaces.rst (hunk 6 above)
Comment (mbauman) [inline doc/manual/interfaces.rst]: Ah, yes, I like sequence much better in general.

File: doc/manual/interfaces.rst (hunk 7 above)
Comment (IainNZ) [inline doc/manual/interfaces.rst]: Hah, I was reading closely, I've been waiting for this document for a long time. Now we just need one for the show/print zoo...

File: doc/manual/interfaces.rst (hunk 2 above)
Comment (mbauman) [inline doc/manual/interfaces.rst]: How about "the type of each item returned by `next`"?  Describing it as the element type is a bit circular.

**Comment by mbauman** [comment]
> Conceptually, what separates these from a "formal" interface?  That the type isn't directly annotated with something that says "Iterable"?

Yes, I think so.  See https://github.com/JuliaLang/julia/issues/6975.  I guess it gets a little blurry for the `AbstractArray` case, since there behaviors are much more defined by it being an `AbstractArray`.  That is, iterables are duck-typed.  AbstractArrays have their own method signature.

**Comment by johnmyleswhite** [comment]
This is really awesome. Well done, @mbauman.

File: doc/manual/interfaces.rst [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+:func:`similar(A) <similar>`                               ``similar(A, eltype(A), size(A))``           Return a mutable array with the same shape and element type
+:func:`similar(A, ::Type{S}) <similar>`                    ``similar(A, S, size(A))``                   Return a mutable array with the same shape and the specified element type
+:func:`similar(A, dims::NTuple{Int}) <similar>`            ``similar(A, eltype(A), dims)``              Return a mutable array with the same element type and the specified dimensions
+:func:`similar(A, ::Type{S}, dims::NTuple{Int}) <similar>` ``Array(S, dims)``                           Return a mutable array with the specified element type and dimensions
+========================================================== ============================================ =======================================================================================
+
+If a type is defined as a subtype of ``AbstractArray``, it inherits a very large set of complicated behaviors including iteration and multidimensional indexing built on top of single-element access.
```
Comment (nalimilan) [inline doc/manual/interfaces.rst]: "Complicated" doesn't make it very appealing. :-) Maybe you could make it more obvious that you *benefit* from "rich" automatic fallbacks by simply defining basic methods?

File: doc/manual/interfaces.rst [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+A key part in defining an ``AbstractArray`` subtype is :func:`Base.linearindexing`. Since indexing is such an important part of an array and often occurs in hot loops, it's important to make both indexing and indexed assignment as efficient as possible.  Array data structures are typically defined in one of two ways: either it's most efficient to access the elements using just one index (using linear indexing) or it intrinsically accesses the elements with indices specified for every dimension.  These two modalities are identified by Julia as ``Base.LinearFast()`` and ``Base.LinearSlow()``.  Converting a linear index to multiple indexing subscripts is typically very expensive, so this provides a traits-based mechanism to enable efficient generic code for all array types.
```
Comment (nalimilan) [inline doc/manual/interfaces.rst]: In "it intrinsically accesses", the "it" sounds a bit weird, as it parallels the previous "it's" which does not refer to anything in particular. Maybe use the passive "or elements are intrinsically accessed"?

File: doc/manual/interfaces.rst [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+.. doctest::
+
+    julia> immutable SquaresVector <: AbstractArray{Int, 1}
+               count::Int
+           end
+           Base.size(S::SquaresVector) = (S.count,)
+           Base.linearindexing(::Type{SquaresVector}) = Base.LinearFast()
+           Base.getindex(S::SquaresVector, i::Int) = i*i;
+
+Note that it's very important to specify the two parameters of the ``AbstractArray``; the first defines the :func:`eltype`, and the second defines the :func:`ndims`.  But that's it takes for our squares type to be an iterable, indexable, and completely functional array:
```
Comment (nalimilan) [inline doc/manual/interfaces.rst]: "that's *what* it takes"?

File: doc/manual/interfaces.rst [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+           Base.similar{T}(A::SparseArray, ::Type{T}, dims::Dims) = SparseArray(T, dims)
+           # Define scalar indexing and indexed assignment up to 3-dimensions
+           Base.getindex{T}(A::SparseArray{T,1}, i1::Int)                   = get(A.data, (i1,), zero(T))
+           Base.getindex{T}(A::SparseArray{T,2}, i1::Int, i2::Int)          = get(A.data, (i1,i2), zero(T))
+           Base.getindex{T}(A::SparseArray{T,3}, i1::Int, i2::Int, i3::Int) =  get(A.data, (i1,i2,i3), zero(T))
+           Base.setindex!{T}(A::SparseArray{T,1}, v, i1::Int)                   = (A.data[(i1,)] = v)
+           Base.setindex!{T}(A::SparseArray{T,2}, v, i1::Int, i2::Int)          = (A.data[(i1,i2)] = v)
+           Base.setindex!{T}(A::SparseArray{T,3}, v, i1::Int, i2::Int, i3::Int) = (A.data[(i1,i2,i3)] = v);
+
+Notice that this is a ``LinearSlow`` array, so we must manually define :func:`getindex` and :func:`setindex!` for each dimensionality we'd like to support.  Unlike the ``SquaresVector``, we are able to define :func:`setindex!`, and so we can mutate the array:
```
Comment (nalimilan) [inline doc/manual/interfaces.rst]: Maybe you should way a word about this difference between ``LinearSlow`` and ``LinearFast`` right after the table above?

File: doc/manual/interfaces.rst [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+Since the ``SparseArray`` is mutable, we were able to override :func:`similar`.  This means that when a base function needs to return an array, it's able to return a new ``SparseArray``:
+
+.. doctest::
+
+    julia> A[1:2,:]
+    2x3 SparseArray{Float64,2}:
+     1.0  4.0  7.0
+     2.0  5.0  8.0
+
+And now, in addition to all the iterable and indexable methods from above, these types can interact with eachother and use all the methods defined in the standard library for ``AbstractArrays``:
```
Comment (nalimilan) [inline doc/manual/interfaces.rst]: "each other"

**Comment by nalimilan** [comment]
Really cool! Don't you think the titles of the sections could be changed to `Iterable`, `Indexable`, and `AbstractArray`s?

File: doc/manual/interfaces.rst (hunk 12 above)
Comment (mbauman) [inline doc/manual/interfaces.rst]: Yes, more can definitely be said.

File: doc/manual/interfaces.rst (hunk 10 above)
Comment (mbauman) [inline doc/manual/interfaces.rst]: Good catch.  This is a bad habit of mine.  See my update — I thought it read nicer to change the first half to agree with the second.

**Comment by nalimilan** [comment]
Thanks, looks good!

**Comment by mbauman** [comment]
Thanks @nalimilan!  Updated (oh, you already found it as I was writing this!).  I like the words iteration and indexing much better than iterable and indexable, but if others think the latter vocabulary is clearer I'll happily change it.

I've also added more cross-links to and from other sections of the manual and standard library.  In doing so, I ran across http://docs.julialang.org/en/latest/manual/arrays/#implementation.  That page and section are written more in a descriptive how-to-use-it fashion instead of the way this is more of a how-to-implement-it.  I think they can both live together nicely, but perhaps some of the information could be consolidated.  Thoughts?

Also, I'm very tempted to document `unsafe_getindex`.

And I really cannot wait for #11242 to get merged.  A syntax is really needed to talk about these things simply and clearly.

**Comment by mbauman** [comment]
I'll merge this now.  It's documentation and there are clamors for hashing/equality, comparison, and display interfaces.  I don't have time to add those myself in the short term, but maybe once there's an obvious place for them to live others will be able to add it.

**Comment by mbauman** [comment]
Shoot, sorry, should have squashed.

**Comment by ScottPJones** [comment]
There is no requirement that an AbstractArray implement a hash function, is there?
The arrays I've implemented in the past frequently didn't use hashing at all (they used a B+tree or a patricia trie).

**Comment by mbauman** [comment]
They get the default implementation for free.  Should be added.

**Pull request merged**